csrc/multidevice/utils.cpp

```diff
@@ -276,36 +277,88 @@ int64_t numDeviceDims(const TensorView* tv) {
 }
 
 namespace {
+std::pair<Val*, bool> computeIndex(
+    IterDomain* id,
+    const std::vector<IterDomain*>& sources,
+    std::unordered_map<IterDomain*, std::pair<Val*, bool>>& id_to_index) {
+  if (id == nullptr) {
+    return {nullptr, false};
+  }
 
-std::vector<IterDomain*> getInputsInTargetDomain(
-    IterDomain* loop_id,
-    const std::vector<IterDomain*>& target_domain) {
-  const std::vector<Val*> inputs_as_vals = IterVisitor::getInputsTo(
-      {loop_id}, {target_domain.begin(), target_domain.end()});
-
-  std::vector<IterDomain*> inputs_as_iter_domains;
-  inputs_as_iter_domains.reserve(inputs_as_vals.size());
-  std::transform(
-      inputs_as_vals.begin(),
-      inputs_as_vals.end(),
-      std::back_inserter(inputs_as_iter_domains),
-      [](Val* val) { return val->as<IterDomain>(); });
-  return inputs_as_iter_domains;
-}
+  std::vector<Expr*> transforms =
+      StmtSort::getExprsBetween({sources.begin(), sources.end()}, {id});
+  for (Expr* transform : transforms) {
+    if (std::all_of(
+            transform->outputs().begin(),
+            transform->outputs().end(),
+            [&](Val* val) {
+              return id_to_index.count(val->as<IterDomain>()) > 0;
+            })) {
+      continue;
+    }
+
+    if (auto* split = dynamic_cast<Split*>(transform)) {
```

Comment:
Shouldn't this compute indices of the input IDs from the output IDs? For example, in the case of producer indexing, `sources` is the logical domain of the producer tensor, and `id` is a loop ID. `getExprsBetween` should return topologically ordered exprs between them. I'd imagine in this case it should traverse the returned list of exprs in the reverse order, computing input indices from output indices.

Reply:
The high-level algorithm here is:
```
for each producer-consumer pair:
  if a producer element and a consumer element that are logical mapped can't be proven to reside on the same GPU:
    return is-resharding

return is-non-resharding
```

Therefore, I chose to start with producer's logical and consumer's root, assign the same index to mapped IterDomains, and compute the loop indices forward. 

I failed to see how backward traversal can be used to implement this algorithm.

Reply:
Oh, I didn't see you set the initial indices of the producer logical and the consumer maybe-root domains. I think this should work for the sake of the resharding analysis, although the name of `computeIndex` is a bit misleading as the computed Vals are not indexing the tensors.

Reply:
> the name of computeIndex is a bit misleading as the computed Vals are not indexing the tensors.

Sure. How about `computeLoopIndex`? The IterDomain* argument is a loop ID.

Reply:
That sounds less confusing.